### ISSUE TITLE
Research Parallelization for Cypress Tests

### ISSUE THREAD

**Issue by holdenhinkle** [department-of-veterans-affairs/va.gov-team#16882]
## Issue Description
Currently Cypress tests take a long time to run (~25 minutes on a local machine). If we setup parallelization we can significantly reduce the time it takes in CI.

Side note: Upgrading Cypress to ~6.0 will probably reduce the time significantly as well.

Cypress parallelization docs - https://docs.cypress.io/guides/guides/parallelization.html

---
## Tasks
- [x] Research how Cypress parallelization could be implemented in our CI pipeline
  * Parallelize by browser?
  * Parallelize by arbitrary test grouping?
  * Split tests based on execution time in order to maximized parallelization effort?
- [x] Research advantages to implementing parallelization for local Cypress test execution
  * Some evidence exists that local execution can be sped up by parallelization on one machine

## Acceptance Criteria
- [x] Existing product space has been surveyed and analyzed
- [x] Research has been collected into a research document
- [x] Testing Tools Team has recommended a specific solution

**Comment by holdenhinkle** [comment]
Researching what 'done' would look like for this ticket.

**Comment by holdenhinkle** [comment]
## Research how Cypress parallelization could be implemented in our CI pipeline
I found four ways:

1. Use **Cypress Dashboard** (PAID) - https://docs.cypress.io/guides/guides/parallelization.html#Overview
2. Use **Knapsack Pro** (PAID) - https://knapsackpro.com/
3. Use **Sorry Cypress** (open source dashboard) - https://sorry-cypress.dev/
3. Roll our own implementation like this - https://testdriven.io/blog/running-cypress-tests-in-parallel/

**Comment by holdenhinkle** [comment]
Working notes for ticket - https://docs.google.com/document/d/1PkvrmBBahtpkl6tOsyD90KC2qdyvrw7SyOYjSUerjIY/edit?usp=sharing

**Comment by holdenhinkle** [comment]
Please note: the following markdown document is missing a few links here and there. Please see [this google doc](https://docs.google.com/document/d/1PkvrmBBahtpkl6tOsyD90KC2qdyvrw7SyOYjSUerjIY/edit#) for original post with all of the links.

# Research how Cypress parallelization could be implemented in our CI pipeline
## Four options:
I discovered four main options in my research:

1. [Cypress Dashboard](https://docs.cypress.io/guides/guides/parallelization.html#Overview) (paid)
2. [Knapsack Pro](https://knapsackpro.com/) (paid)
3. [Sorry Cypress](https://sorry-cypress.dev/) (free, open-source dashboard)
4. Roll our own implementation like [this](https://testdriven.io/blog/running-cypress-tests-in-parallel/)

## Parallelize by browser?
### 1. Cypress Dashboard (paid)
Yes.

Group by browser - https://docs.cypress.io/guides/guides/parallelization.html#Grouping-by-browser

Also, see Linking CI machines for parallelization or grouping - https://docs.cypress.io/guides/guides/parallelization.html#Grouping-by-spec-context 

--ci-build id - https://docs.cypress.io/guides/guides/command-line.html#cypress-run-ci-build-id-lt-id-gt 
### 2. Knapsack Pro (paid)
Yes.

According to the FAQ - https://github.com/KnapsackPro/knapsack-pro-cypress#knapsack-pro-core-features-faq

You can pass many of Cypress CLI arguments to @knapsack-pro/cypress. We need to verify that we can pass `--browser` and `--group` options.
### 3. Sorry Cypress (free, open-source)
Yes.

Source - Cypress Parallelization on Jenkins using Sorry-Cypress | by Aditya Trivedi 

To group your tests you can change the CI build command to below, so your test will appear as “master-1”, “master-2”:

```
sh "npm run cy:run:parallel -- --ci-build-id ${env.BRANCH_NAME}-${env.BUILD_ID}"
```

`cypress run --ci-build-id` BUILD_NUMBER documentation - https://docs.cypress.io/guides/guides/command-line.html#cypress-run-ci-build-id-lt-id-gt 
### 4. Roll our own using example implementation
Yes, but bad implementation.

The run command is hard-coded into in the js file that generates part of the `.circleci/config.yml` file, which is not optimal:

```
{
  run: {
    name: `Running cypress tests ${index + 1}`,
    command: `$(npm bin)/cypress run --spec cypress/integration/${value}`,
  },
},
```

## Parallelize by arbitrary test grouping?
### 1. Cypress Dashboard (paid)
Yes.

Grouping to label parallelization - https://docs.cypress.io/guides/guides/parallelization.html#Grouping-to-label-parallelization

Additional documentation on groups - https://docs.cypress.io/guides/guides/command-line.html#cypress-run-group-lt-name-gt 

You can also group by spec - 
https://docs.cypress.io/guides/guides/parallelization.html#Grouping-by-spec-context
### 2. Knapsack Pro (paid)
Probably.

According to the FAQ - https://github.com/KnapsackPro/knapsack-pro-cypress#knapsack-pro-core-features-faq

You can pass many of Cypress CLI arguments to @knapsack-pro/cypress. We need to verify that we can pass `--browser` and `--group` options.
### 3. Sorry Cypress (free, open-source)
It appears you can.

Here’s a local example - https://github.com/sorry-cypress/sorry-cypress/tree/master/example

You can pass in the standard Cypress run options:
npx cypress run --record --key whatever --parallel --ci-build-id local-01

I can spend more time investigating if required.
### 4.  Roll our own using example implementation
No.

The referenced implementation does not allow grouping. It simply looks for all Cypress spec files:

```
// get spec files as an array
const files = fs.readdirSync(path.join(__dirname, '..', 'cypress', 'integration')).filter(fn => fn.endsWith('.spec.js'));
```

Also, this implementation doesn’t use a queue. Each spec is added iteratively to a group, one for each machine:

```
function createJSON(fileArray, data) {
  for (const [index, value] of fileArray.entries()) {

    // omitted code

    data.workflows.build_and_test.jobs.push({
        [`test${index + 1}`]: {
          requires: [
            'build',
          ],
        },
      });
    }
  return data;
}

/ get spec files as an array
const files = fs.readdirSync(path.join(__dirname, '..', 'cypress', 'integration')).filter(fn => fn.endsWith('.spec.js'));
// read circle.json
const circleConfigJSON = require(path.join(__dirname, 'circle.json'));
// add cypress specs to object as test jobs
const data = createJSON(files, circleConfigJSON);
// write file to disc
writeFile(data);
```

## Split tests based on execution time in order to maximized parallelization effort?
### 1. Use Cypress Dashboard (paid)
Yes.

It does load balancing automatically.
### 2. Use Knapsack Pro (paid)
Yes.

Knapsack Pro implements a queue that hands off a test to a machine when a machine becomes available. 

Knapsack Pro calls this mode Queue Mode. Here’s a video about it - Why and how to split Ruby tests across CI nodes in a dynamic way using Knapsack Pro Queue Mode 

Even though this video, which was created in 2018, says this ‘mode’ is for Ruby tests (RSpec and Minitest), it does apply to Cypress tests:

Source
@knapsack-pro/cypress runs your E2E tests with Cypress.io test runner and does dynamic tests allocation across parallel CI nodes using KnapsackPro.com Queue Mode to provide the fastest CI build time (optimal test suite timing).

Learn about Knapsack Pro Queue Mode in the video how to run tests with dynamic test suite split and learn what CI problems can be solved thanks to it.
### 3. Use Sorry Cypress (free, open-source)
Unknown.

Cannot quickly find information about load balancing.

I can spend more time investigating if required.
### 4.  Roll our own using example implementation
No.
## Research advantages to implementing parallelization for local Cypress test execution

> “Technically, it's possible. Cypress does not recommend it since running multiple cypress instances on the same machine consumes lots of resources (CPU over all) and it slows down the entire machine performances with useless results.”

=> Source (not official source)
### Options for running local tests in parallel:
#### cypress-parallel
https://github.com/tnicola/cypress-parallel

Reduce up to 40% your Cypress suite execution time parallelizing the test run on the same machine.
#### cypress-parallel-specs-locally
https://github.com/Shelex/cypress-parallel-specs-locally
#### BrowserStack - Cypress testing is in beta
Run Cypress Tests in Parallel
### Pros and Cons
#### Cypress Dashboard
**Pros:**

- Parallelization is easy to setup with CircleCi
- Intelligent Load Balancing of parallelized tests
- Lots of other great, useful features not related to this ticket

**Cons:**
Might be expensive
#### Knapsack Pro
**Pros:**

- Parallelization uses a queue (Queue Mode) for load balancing
- Can also be used for Rspec tests (and Jest, Minitest, Test::Unit, Cucumber, Spinach, Turnip)

**Cons:**

- Pricing is difficult to understand and might be expensive. We might as well go with Cypress.
- Possible con: When adding a new application to vets-website, or a new directory for Cypress tests, you have to add the new directory to the KNAPSACK_PRO_TEST_FILE_PATTERN environment variable

#### Sorry Cypress
**Pros:**

- Sorry-cypress is an open-source, on-premise, self-hosted alternative to Cypress dashboard (AKA ‘free’)
- Is open-source and is actively maintained/under development
- run cypress tests in parallel with no limitations
- upload screenshots and videos to your own storage
- browse test results, screenshots and video recordings
- self-hosted - use your own infrastructure, own your data
- integrate with GitHub or anything else via webhooks
- ready to use Docker images, Kubernetes and AWS Cloud formation template
- can contribute to open-source

**Cons:**

- We’d have to implement it in CircleCi ourselves (there’s no documentation for that)
- Dashboard is not as polished or as feature-rich as the Cypress Dashboard
- From article: One thing that may not work with parallelization is the ability to report code coverage.

#### Roll our own using example implementation
**Pros:**

- Example implementation can be used for inspiration, if nothing else

**Cons:**

- The man-hours it takes to build and maintain a custom solution
## Bonus: Resources for Setting-up CircleCI with Cypress Dashboard (Cypress subscription required)
Recording test results and spec parallelization requires Cypress Dashboard subscription.

**Resources:**

- CircleCi and Cypress Orb: https://github.com/cypress-io/circleci-orb-parallel-example
- circle.yml example - https://github.com/cypress-io/circleci-orb-parallel-example/blob/master/circle.yml 
- Cypress Orb - https://github.com/cypress-io/circleci-orb 
- Configure circle.yml to run groups of tests - https://github.com/cypress-io/circleci-orb/blob/master/docs/examples.md#groups 
- Configure circle.yml to fun tests on a certain browser - https://github.com/cypress-io/circleci-orb/blob/master/docs/examples.md#chrome

**Comment by pjhill** [comment]
Information converted to a Confluence document [here](https://vfs.atlassian.net/wiki/spaces/VI/pages/913702932/Research+Setup+Parallelization+for+Cypress+Tests#cypress-dashboard-paid).

**Comment by pjhill** [comment]
The research phase of this initiative has been completed. We will work the recommendation and procurement of a relevant product under a new ticket.
